docs/internal/generics.md

```diff
@@ -0,0 +1,114 @@
+
+// This creates another concrete type, now with a more restrictive argument
+let y : MyGenericType[F64, MyClass] = MyGenericType;
+
+y.field = 42; // ERROR: No known conversion between integer literals and float types
+
+y.method(c); // OK: This is the expected type
+y.method(o); // ERROR: Even though MyOtherType implements the same interface, it's not he same type
+```
+
+In the example above, there is one generic class called `MyGenericClass` and two implementations (or specialisations) with types `[U32, Hashable]` and `[F64, MyClass]`.
+If all types are used correctly (ex. removing the ERROR lines above), the compiler will create two concrete classes from the one generic class: one for each specialisation.
+
+## Type Parameters
+
+Type parameters are the named types that can be used inside a generic declaration in place of actual types.
+In the example above, the generic class `MyGenericClass` has two type parameters: `MyType` and `Hashable`.
+Those names are used inside the class in place of existing (declared) types.
+In each specialisation, the type arguments (ex. `[U32, Hashable]`) replace the named arguments and the specialisation types are matched 1:1.
+
+Type parameters have the following properties:
+* A name: The name that will be used in the generic class as a substitute for the specialisation's passed declared name.
+* A restriction: What are the requirements (ex. interfaces, super-types, capabilities) of the type being passed in each specialisation.
+* A default type: In case the type is omitted from the specialisation declaration, a default type is assumed.
+
+The syntax is: `[Name : Restriction = Default Value]`.
+
+On the example above, only the names were provided.
+Even though the names were `MyType` and `SomeInterface`, they could have been initialised with anything, included concrete types, interfaces and capabilities.
+It's up to the compiler's type checks to make sure the types passed by each specialisation make sense.
+For example, a field with just a capability type cannot be properly initialised, so that's a type error.
+
+Example:
+```ts
+// A can be any type with any capability
+// B can be any type, but must have a mut capability
+// C can also be any type, but if none is passed, it's U32
+class Generic[A, B : mut, C = U32] { ... }
+```
+
+In the example above, `C` is a dangerous declaration.
+Having a concrete default type on a unrestricted generic type may lead the implementation to assume users will only pass numeric types.
+
+To correct that mistake, a complete declaration would help:
+```ts
+class Generic[A, B : mut, C : (U8 | U16 | U32 | U64) & imm = U32 & imm] { ... }
+```
+
+Now we can see that the programmer only expected unsigned integers, which was even more restricted than previously thought.
+To make this easier to read, one can use type aliases:
+
+```ts
+type UInts = (U8 | U16 | U32 | U64) & imm;
+...
+class Generic[A, B : mut, C : UInts = U32 & imm] { ... }
+```
```

Comment:
Same subtyping question as above.  I don't think this means what you think it means.  In the trivial case, it would be allowable to instantiate this with `UInts` as the generic parameter and now you have dynamic dispatch in a place you presumably don't want it (you wouldn't restrict yourself to fixed-size integers if you didn't want arithmetic to be cheap and / or storage space to be small and well-defined).

Reply:
Is there a way to say: "only one of those" rather than "any combination of those"? Do we even want that?

I imagine @sylvanc would say: Well, if programmer pass a union type in the end, they'll get dynamic dispatch and it's their problem. If they want concrete behaviour they should have passed concrete types to begin with.